### PR TITLE
GLT-535 Save detailed samples sequentially

### PR DESCRIPTION
**Pull request by tonydebat**
Added a serial function to execute an array of functions in order. Used this to bulk samples.

### CODE REVIEW

File: miso-web/src/main/webapp/scripts/sample_hot.js [hunk 1]
```diff
@@ -914,6 +915,17 @@ Sample.hot = {
         // send it through the parser to get a sampleData array that isn't merely a reference to Sample.hot.hotTable.getSourceData()
         var sampleData = JSON.parse(JSON.parse(JSON.stringify(Sample.hot.hotTable.getSourceData())));
         
+        // Array of save functions, one for each line in the table.
+        var sampleSaveArray = new Array();
+
```
Comment (callunity) [inline miso-web/src/main/webapp/scripts/sample_hot.js]: From what I understand, it's more idiomatic in JavaScript to use ``var sampleSaveArray = [];``

Comment (tonydebat) [inline miso-web/src/main/webapp/scripts/sample_hot.js]: Updated.

File: miso-web/src/main/webapp/scripts/sample_hot.js [hunk 2]
```diff
@@ -793,10 +793,11 @@ Sample.hot = {
     return sample;
   },
   
-  saveOneSample: function (data, index, messages, numberToSave) {
+  saveOneSample: function (data, index, messages, numberToSave, callback) {
     var xhr = new XMLHttpRequest();
     xhr.onreadystatechange = function() {
       if (xhr.readyState === XMLHttpRequest.DONE) {
+    	callback(); // Indicate request has completed.
```
Comment (apmasell) [inline miso-web/src/main/webapp/scripts/sample_hot.js]: Is this guaranteed to execute in the desired order? Should the callback be invoked after the save function?

File: miso-web/src/main/webapp/scripts/sample_hot.js [hunk 3]
```diff
@@ -934,6 +947,22 @@ Sample.hot = {
     });
   },
   
+  /**
+   * Serial execution of an array of functions. 
+   * @param {Function[]} arrayOfFunctions - Each function takes a single callback function parameter.
+   */
+  serial: function(arrayOfFunctions) {
+	  index = 0;
```
Comment (apmasell) [inline miso-web/src/main/webapp/scripts/sample_hot.js]: Missing `var`.

File: miso-web/src/main/webapp/scripts/sample_hot.js (hunk 2 above)
Comment (callunity) [inline miso-web/src/main/webapp/scripts/sample_hot.js]: The callback starts sending the second sample to the server while the front end processes whether the first sample was successfully saved, so having those execute out of order shouldn't be a problem.

**Comment by callunity** [comment]
:+1:

File: miso-web/src/main/webapp/scripts/sample_hot.js [hunk 4]
```diff
@@ -934,6 +947,22 @@ Sample.hot = {
...
+	  
+	  function serialCallBack() {
+		  if(index < (arrayOfFunctions.length)) {
+		     arrayOfFunctions[index++](serialCallBack);
```
Comment (apmasell) [inline miso-web/src/main/webapp/scripts/sample_hot.js]: I find this mutation of state to be really difficult to reason about. Can this be something like:

```
fuction(aof) {
   var invokeNext = function(index) {
       if (index < aof.length) { aof[index](function() { invokeNext(index + 1);} ); }
   }
   invokeNext(0);
}

File: miso-web/src/main/webapp/scripts/sample_hot.js (hunk 2 above)
Comment (tonydebat) [inline miso-web/src/main/webapp/scripts/sample_hot.js]: Yes. At this point the REST request is complete and either ended in success or failure. The next save can be triggered before the successSave or failSave callbacks update the page with the results.

**Comment by apmasell** [comment]
:+1:

**Pull request merged**